Code assist for Anonymous class creation is broken

snjeza opened the pull request:
Fixes #57 

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

fbricon commented:
I can see at least 3 issues:
- indentation is off by 2 tabs, last bracket not aligned with the opening one
- method bodies should not be placeholders, else you're block editing all of them as soon as you type
- Depending on when you apply the textedit, it is not cleanly injected:
![sep-05-2017 09-18-38](https://user-images.githubusercontent.com/148698/30063089-64ef4372-921b-11e7-8ef6-73e3e4fee982.gif)

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/AnonymousTypeCompletionProposal.java:
```diff
@@ -0,0 +1,260 @@
+		IRegion lineInfo = document.getLineInformationOfOffset(fReplacementOffset);
+		int indent = Strings.computeIndentUnits(document.get(lineInfo.getOffset(), lineInfo.getLength()), project);
+		Map<String, String> options = project != null ? project.getOptions(true) : JavaCore.getOptions();
+		options.put(DefaultCodeFormatterConstants.FORMATTER_INDENT_EMPTY_LINES, DefaultCodeFormatterConstants.TRUE);
```
No, why override the project options?

fbricon commented:
See 
![sep-05-2017 11-34-47](https://user-images.githubusercontent.com/148698/30069273-4d91a6da-922e-11e7-834b-8a377f6c2bf4.gif)

fbricon commented on org.eclipse.jdt.ls.target/org.eclipse.jdt.ls.tp.target:
```diff
@@ -18,7 +18,7 @@
 </location>
 <location includeAllPlatforms="false" includeConfigurePhase="false" includeMode="planner" includeSource="true" type="InstallableUnit">
-<unit id="org.eclipse.lsp4j.sdk.feature.group" version="0.3.0.v20170830-0917"/>
+<unit id="org.eclipse.lsp4j.sdk.feature.group" version="0.3.0.v20170905-1253"/>
```
was already fixed in master, you can drop that commit

fbricon commented:
Completion is broken when you type the opening parenthesis:
![sep-05-2017 12-23-27](https://user-images.githubusercontent.com/148698/30071589-3d21302a-9235-11e7-8f44-13c7f176b128.gif)

snjeza commented:
> Can we keep snippetSupport if the generated body contains only 1 method?

Fixed.

> Completion is broken when you type the opening parenthesis:

Fixed.

fbricon commented:
>> Completion is broken when you type the opening parenthesis:

> Fixed.

Nope, I still see the issue with the latest changes. 

Can you please add unit tests for the use cases you fix.

snjeza commented:
> Nope, I still see the issue with the latest changes.

I can reproduce the issue. It happens when I type fast. Trying to fix it.

snjeza commented:
I have added content assist for an anonymous class declaration. In order to test, you need to call Ctrl-Space on
```
new Runnable(|)
```

However, I haven't succeeded to solve the issue described in https://github.com/eclipse/eclipse.jdt.ls/pull/343#issuecomment-327229439
The issue is caused by the way in which VS Code controls close brackets.
When you type 

```
new Runnable|
```

and enter open parentheses - (

VS Code calls completion and adds close parentheses - ). If completion is executed before adding close parentheses, VS Code will unnecessarily add ) and we will get an invalid code.
The issue isn't reproducible when setting the editor.autoClosingBrackets property to false:

```
"editor.autoClosingBrackets": false
```

Not sure how to fix this issue, i.e., if there is a way to ensure a completion action to be called after closing parentheses action.

snjeza commented:
See https://ci.eclipse.org/ls/job/jdt-ls-pr/290/testReport/junit/org.eclipse.jdt.ls.core.internal.correction/JavadocQuickFixTest/testInvalidQualification1/

fbricon commented:
@aeschli any idea/advice about the bracket issue?

fbricon commented:
retest this please

fbricon commented:
So this doesn't enable snippets only when there's 1 method, but on the 1st method found, which is probably better.

But it turns out this enables snippets on the last method of the class. So you rather want to add the snippet if ` i == methodsToOverride.length-1`

Also please, add more unit tests about the bugs you fixed (indentation, snippet support, opening parenthesis)

snjeza commented:
> `i == methodsToOverride.length-1`

Fixed.

> Also please, add more unit tests about the bugs you fixed (indentation, snippet support, opening parenthesis)

The existing tests already test indentation and snippet support because they check "expected text".
I am going to add a test that will use a type with more methods.
I haven't fixed anything related to parenthesis. Suppose it is an upstream issue.

fbricon commented on org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/CompletionProposalReplacementProvider.java:
```diff
@@ -166,6 +168,55 @@ public void updateReplacement(CompletionProposal proposal, CompletionItem item,
+				if (document.getChar(offset) == '(') {
+					pos = offset;
+				}
+				if (document.getChar(offset + 1) == '(') {
```
I can't tell if it should test for `)` or it's `(` a 2nd time, on purpose. Either fix it or document it please.

snjeza replied:
It's correct. I have added a comment.

fbricon replied:
if offset ==  document.getLength()-1, then document.getChar(offset + 1) will throw a BadLocationException.
Although it's unlikely to happen, it still needs to be guarded against

fbricon replied:
it should be 
```java
if (document.getChar(offset + 1) == '(') {
   pos = offset + 1;
} else if (document.getChar(offset) == '(') {
   pos = offset;
]
```

But it's

fbricon replied:
I'm still having a hard time reconciling this piece of code from org.eclipse.jdt.internal.ui.text.java.AnonymousTypeCompletionProposal 
```java
if (pos < document.getLength() && document.getChar(pos) == ')') {
   int currentLength= getReplacementLength();
   if (replacementStringEndsWithParentheses)
     setReplacementLength(currentLength + pos - offset);
   else
     setReplacementLength(currentLength + pos - offset + 1);
}
```
I don't see your comment

snjeza commented:
There are some more issues. If you call Ctrl-Space on

```
new Runnable(  |  )
```

or

```
public class Test {
	public void test() {
		run("name", new Runnable(|, 1);
	}

	void run(String name, Runnable runnable, int i) {
	}
}
```

you will get invalid an code. Trying to fix the issues.

fbricon commented:
nice catch, please add unit tests for those cases too

snjeza commented:
The following case will create an incorrect code in Eclipse:

```
public class Test {
	public void test() {
		run("name", new Runnable|
		, 1);
	}
	void run(String name, Runnable runnable, int i) {

	}
}
```

and the following will throw an exception in Eclipse:
```
public class Test {
	public void test() {
		x = new Runnable(|
```
The left parenthesis is the end of the file.

fbricon commented:
Ok, we have at least the functionality working in most use cases, let's start with that. Nice work Snjeza.
